File: erlang/graphics.erl
-module(graphics).
-export([clear_screen/0, print_game/4]).

-define(BLACK, 40).
-define(WHITE, 47).


odd_number(N)        -> 2 * N + 1.
% Terminal functions 
clear_screen()       -> io:format("\e[2J\e[H").
cursor_up(N)         -> io:format("\e[~wA", [N]).
cursor_down(N)       -> io:format("\e[~wB", [N]).
cursor_right(N)      -> io:format("\e[~wC", [N]).
cursor_left(N)       -> io:format("\e[~wD", [N]).
cursor_restore(Size) -> cursor_down(1), cursor_left(odd_number(Size)).


select_color(Disc) when Disc == ?BLACK -> print_color(?BLACK);
select_color(Disc) when Disc == ?WHITE -> print_color(?WHITE);
select_color(Disc) -> print_color((Disc rem 6) + 1 + 40).


print_color(Color) -> io:format("\e[~wm", [Color]).
print_color(Color, Size) ->
    select_color(Color),
    lists:map(fun(_) -> io:format(" ") end, lists:seq(1, Size)).


print_disc(Disc, Size) when Disc == 0 ->
    EmptySpace = odd_number(Size) div 2,
    print_color(?BLACK, EmptySpace),
    print_color(?WHITE, 1),
    print_color(?BLACK, EmptySpace),
    cursor_restore(Size);

print_disc(Disc, Size) ->
    DiscSize = odd_number(Disc),
    SpaceCount = (odd_number(Size) - DiscSize) div 2,
    print_color(?BLACK, SpaceCount),
    print_color(Disc, DiscSize),
    print_color(?BLACK, SpaceCount),
    cursor_restore(Size).


print_tower(Size, Tower) ->
    FreeSpaces = lists:seq(1, Size - length(Tower)),
    lists:map(fun(_) -> print_disc(0, Size) end, FreeSpaces),
    lists:map(fun(X) -> print_disc(X, Size) end, Tower),
    % Undo last cursor movement
    cursor_up(1),
    cursor_right(odd_number(Size)),
    % Prepare for next tower
    cursor_right(1),
    cursor_up(Size - 1).


print_tower_id(Size, ID) when ID =< 3 ->
    DiscSize = odd_number(Size),
    Center = (DiscSize - 1) div 2,
    lists:map(
        fun(X) ->
            case X == Center of 
                true -> io:format("~w", [ID]);
                false -> io:format(" ")
            end
        end,
        lists:seq(0, DiscSize - 1)
    ),
    io:format(" "),
    print_tower_id(Size, ID + 1);

print_tower_id(_, _) -> ok. 

print_hud(Size) -> 
    print_tower_id(Size, 1),
    io:format("\n\nPossible moves: 12, 13, 21, 23, 31 and 32."),
    io:format("\nInsert 0 to start a new game."),
    io:format("\nInsert a move: ").


print_game(Size, TowerA, TowerB, TowerC) ->
    clear_screen(),
    print_tower(Size, TowerA),
    print_tower(Size, TowerB),
    print_tower(Size, TowerC),
    % Print HUD
    cursor_down(Size),
    cursor_left(100),
    print_hud(Size).

File: haskell/app/Graphics.hs
module Graphics (clearScreen, printGame) where

black :: Int
black = 40

white :: Int
white = 47

oddNumber :: Int -> Int
oddNumber n = 2 * n + 1

clearScreen :: IO ()
clearScreen = putStr "\x1b[2J\x1b[H"

cursorUp :: Int -> IO ()
cursorUp n = putStr ("\x1b[" ++ show n ++ "A")

cursorDown :: Int -> IO ()
cursorDown n = putStr ("\x1b[" ++ show n ++ "B")

cursorRight :: Int -> IO ()
cursorRight n = putStr ("\x1b[" ++ show n ++ "C")

cursorLeft :: Int -> IO ()
cursorLeft n = putStr ("\x1b[" ++ show n ++ "D")

cursorRestore :: Int -> IO ()
cursorRestore size = do
    cursorDown 1
    cursorLeft (oddNumber size)

selectColor :: Int -> IO ()
selectColor disc
    | disc == black = printColor black
    | disc == white = printColor white
    | otherwise = printColor ((disc `mod` 6) + 1 + 40)

printColor :: Int -> IO ()
printColor color = putStr ("\x1b[" ++ show color ++ "m")

printLine :: Int -> Int -> IO ()
printLine color size = do
    selectColor color
    mapM_ (\_ -> putStr " ") [1..size]

printDisc :: Int -> Int -> IO ()
printDisc 0 size = do
    let emptySpace = oddNumber size `div` 2
    printLine black emptySpace
    printLine white 1
    printLine black emptySpace
    cursorRestore size

printDisc disc size = do
    let discSize = oddNumber disc
    let spaceCount = (oddNumber size - discSize) `div` 2
    printLine black spaceCount
    printLine disc discSize
    printLine black spaceCount
    cursorRestore size

printTower :: Int -> [Int] -> IO ()
printTower size tower = do
    let freeSpaces = [1..(size - length tower)]
    mapM_ (\_ -> printDisc 0 size) freeSpaces
    mapM_ (`printDisc` size) tower
    -- Undo last cursor movement
    cursorUp 1
    cursorRight (oddNumber size)
    -- Prepare for next tower
    cursorRight 1
    cursorUp (size - 1)

printTowerId :: Int -> Int -> IO ()
printTowerId size index
    | index > 3 = return ()
    | otherwise = do
        let discSize = oddNumber size
        let center = (discSize - 1) `div` 2
        mapM_ (\x ->
            if x == center
                then putStr (show index)
                else putStr " "
            ) [0..(discSize - 1)]
        putStr " "
        printTowerId size (index + 1)

printHud :: Int -> IO ()
printHud size = do
    printTowerId size 1
    putStr "\n\nPossible moves: 12, 13, 21, 23, 31 and 32."
    putStr "\nInsert 0 to start a new game."
    putStr "\nInsert a move: "

printGame :: Int -> [Int] -> [Int] -> [Int] -> IO ()
printGame size towerA towerB towerC = do
    clearScreen
    printTower size towerA
    printTower size towerB
    printTower size towerC
    -- Print HUD
    cursorDown size
    cursorLeft 100
    printHud size

